onnxruntime/core/providers/cpu/tensor/utils.h

```diff
@@ -177,22 +177,52 @@ struct SliceIteratorBase {
 
   // Initialize initial skip and inner_extent.
   void Init(gsl::span<const int64_t> dims, gsl::span<const int64_t> starts, gsl::span<const int64_t> steps) {
-    ORT_ENFORCE(dims.size() == starts.size() &&
-                dims.size() == extents_.size() &&
-                dims.size() >= steps.size());
+    const auto dims_size = dims.size();
+    ORT_ENFORCE(dims_size == starts.size() &&
+                dims_size == extents_.size() &&
+                dims_size >= steps.size());
 
     SafeInt<size_t> pitch = 1;
     // Initial skip, so that input_ points to the first element to copy
-    for (size_t i = dims.size(); i-- > 0;) {
+    for (size_t i = dims_size; i-- > 0;) {
       input_ += pitch * starts[i] * element_size_;
       pitch *= static_cast<size_t>(dims[i]);
     }
 
-    inner_extent_ = static_cast<size_t>(extents_[dims.size() - 1]);
-    //It could be -1
-    inner_step_ = static_cast<ptrdiff_t>(dims.size() == steps.size()
-                                             ? steps[dims.size() - 1]
+    inner_extent_ = static_cast<size_t>(extents_[dims_size - 1]);
+    // It could be -1
+    const auto steps_size = steps.size();
+    inner_step_ = static_cast<ptrdiff_t>(dims_size == steps_size
+                                             ? steps[steps_size - 1]
                                              : 1);
+
+    // This block serves to maximize the amount of data that is copied
+    // in one shot when trailing axis are not being sliced and, therefore,
+    // more data can be copied in one shot. The inner dimension with step = 1
+    // can always copy with more than one element unless its extent is 1.
+    // However, when more than one trailing dimensions have steps one and copied
+    // in its entirety, we can copy bigger blocks at the same time and start skipping
+    // them as the skips are expected to be zero then for them.
+    SafeInt<int64_t> max_copyable_elements = (inner_step_ == 1) ? inner_extent_ : 1;
+    last_batching_axis_ = dims_size - 1;
+
+    // Check if inner dimension is copied as a block in its entirety
+    if (dims_size > 1 && inner_step_ == 1 && inner_extent_ == gsl::narrow<size_t>(dims[dims_size - 1])) {
+      for (size_t dim = dims_size - 2;; dim--) {
+        if (dim < steps_size && steps[dim] != 1) {
+          break;
+        }
+        max_copyable_elements *= extents_[dim];
```

Comment:
Is there code coverage of this from the existing unit tests? Not sure if you validated with those or a specific model.

Reply:
Yes, Slice now takes advantage of the new interface and I have validated it with the customer model. The new code is covered.

Reply:
Is it covered by a unit test I can look at? e.g. one that has a similar inputs to the Slice in the customer model.

Wondering if it would make sense to store a list of block sizes instead of having a single max_copying_elements_block_ value - but I don't have an example of the Slice params you were testing with to figure out if that would be of value.

Reply:
This is a good example of coverage. https://github.com/microsoft/onnxruntime/blob/c5908c4ea273312deefa6181c17f6d8bab71c155/onnxruntime/test/providers/cpu/tensor/slice_op.test.cc#L636

Reply:
List of copyable block sizes is stored in extents_

Reply:
Ignore the part about storing a list - that would only be useful if an axis could be repeated in `axes` which isn't allowed.

Reply:
Wouldn't Slice5D_LargeStep behave the same without your change as it's slicing on axis 1 and only taking 1 element, so it's the same amount of data that flattening the last 3 dims would have picked. 

If so could we add a test where the new code is a) used for multiple slices, and b) the block size would differ from what the existing code does?

Reply:
Added test `SliceTest.Slice5D_CopyAxis2LargeBlock` that would copy blocks by 8, not by 4, but still does it in multiple slices because of axis 1 slicing.